File: src/main/java/com/github/nsc/de/shake/parser/node/expression/SubNode.java
package com.github.nsc.de.shake.parser.node.expression;

import com.github.nsc.de.shake.parser.node.ValuedNode;
import com.github.nsc.de.shake.parser.node.factor.DoubleNode;
import com.github.nsc.de.shake.parser.node.factor.IntegerNode;

public class SubNode extends ExpressionNode {
    public SubNode(ValuedNode left, ValuedNode right, int operatorPosition) {
        super(left, right, operatorPosition);
    }
    public SubNode(int left, ValuedNode right, int operatorPosition) {
        this(new IntegerNode(left), right, operatorPosition);
    }
    public SubNode(ValuedNode left, int right, int operatorPosition) {
        this(left, new IntegerNode(right), operatorPosition);
    }
    public SubNode(int left, int right, int operatorPosition) {
        this(new IntegerNode(left), new IntegerNode(right), operatorPosition);
    }
    public SubNode(double left, ValuedNode right, int operatorPosition) {
        this(new DoubleNode(left), right, operatorPosition);
    }
    public SubNode(ValuedNode left, double right, int operatorPosition) {
        this(left, new DoubleNode(right), operatorPosition);
    }
    public SubNode(double left, double right, int operatorPosition) {
        this(new DoubleNode(left), new DoubleNode(right), operatorPosition);
    }

    @Override
    public char getOperator() {
        return '-';
    }
}

File: parser/src/main/kotlin/com/github/nsc/de/shake/parser/node/expression/SubNode.kt
package com.github.nsc.de.shake.parser.node.expression

import com.github.nsc.de.shake.parser.node.ValuedNode
import com.github.nsc.de.shake.parser.node.factor.DoubleNode
import com.github.nsc.de.shake.parser.node.factor.IntegerNode
import com.github.nsc.de.shake.util.characterinput.position.PositionMap

@Suppress("unused")
class SubNode(map: PositionMap, left: ValuedNode, right: ValuedNode, operatorPosition: Int) :
    ExpressionNode(map, left, right, operatorPosition) {
    constructor(map: PositionMap, left: Int, right: ValuedNode, operatorPosition: Int) : this(
        map,
        IntegerNode(map, left),
        right,
        operatorPosition
    )

    constructor(map: PositionMap, left: ValuedNode, right: Int, operatorPosition: Int) : this(
        map,
        left,
        IntegerNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Int, right: Int, operatorPosition: Int) : this(
        map,
        IntegerNode(map, left),
        IntegerNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Double, right: ValuedNode, operatorPosition: Int) : this(
        map,
        DoubleNode(map, left),
        right,
        operatorPosition
    )

    constructor(map: PositionMap, left: ValuedNode, right: Double, operatorPosition: Int) : this(
        map,
        left,
        DoubleNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Double, right: Double, operatorPosition: Int) : this(
        map,
        DoubleNode(map, left),
        DoubleNode(map, right),
        operatorPosition
    )

    override val operator: Char
        get() = '-'

    override fun toJson(): Map<String, *> = mapOf("name" to "SubNode", "left" to left, "right" to right)
}